### PR TITLE
Throw if two inner_hits have the same name

### LINKED ISSUE #37584: Inner_hits in should clauses override each other
This issues is somewhat similar to #16218, but it manifests itself if two `should` clauses contain `has_child` with `inner_hits` and different queries and only one of the query in the last clause doesn't matches the document returned by the first clause.

This is a repro for 6.x

```
DELETE test
PUT test
{
  "mappings": {
    "_doc": {
      "properties": {
        "my_join_field": {
          "type": "join",
          "relations": {
            "question": "answer"
          }
        },
        "text": {
          "type": "text"
        }
      }
    }
  }
}

POST test/_doc/_bulk
{"index":{"_id": "1"}}
{"text": "This is a question","my_join_field": {"name": "question"}}
{"index":{"_id": "2"}}
{"text": "This is another question","my_join_field": {"name": "question"}}
{"index":{"_id": "3", "_routing": "1"}}
{"text": "This is an answer", "my_join_field": {"name": "answer","parent":"1"}}
{"index":{"_id": "4", "_routing": "1"}}
{"text": "This is another answer", "my_join_field": {"name": "answer","parent":"1"}}

POST test/_search

# Doesn't return inner hits
POST test/_search
{
  "query": {
    "bool": {
      "should": [
        {
          "has_child": {
            "type": "answer",
            "query": {
              "match": {
                "text": "answer"
              }
            },
            "inner_hits": {}
          }
        },
        {
          "has_child": {
            "type": "answer",
            "query": {
              "match": {
                "text": "noanswer"
              }
            },
            "inner_hits": {}
          }
        }
      ]
    }
  }
}

# Returns inner hits:
POST test/_search
{
  "query": {
    "bool": {
      "should": [
        {
          "has_child": {
            "type": "answer",
            "query": {
              "match": {
                "text": "noanswer"
              }
            },
            "inner_hits": {}
          }
        },
        {
          "has_child": {
            "type": "answer",
            "query": {
              "match": {
                "text": "answer"
              }
            },
            "inner_hits": {}
          }
        }
      ]
    }
  }
}
``` 
Repro for 7.x can be found [here](https://gist.github.com/imotov/53c2a81b7ba0db2663a50a3f6d8fa7e3)

Originally reported in https://discuss.elastic.co/t/inner-hits-has-child/164627/10

### PR DESCRIPTION
**Pull request by dvehar**
Fixes #37584

### CODE REVIEW

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @dvehar , the change looks great.
I left some comments regarding the unit tests, we also need an integration test that check that setting two inner hits with the same name throws an exception during the execution of the query. Can you add one in `NestedIT` for instance ?

**Comment by dvehar** [comment]
I'll work on the integration test soon

**Comment by dvehar** [comment]
@jimczi Could you provide some guidance as to what my test case should look like?

```
public void testExtractInnerHitBuildersWithDuplicate() throws Exception {
        assertAcked(
            prepareCreate("idxduplicatehitnames")
                .setSettings(Settings.builder().put(SETTING_NUMBER_OF_SHARDS, 1).put(SETTING_NUMBER_OF_REPLICAS, 0))
                .addMapping("product", "categories", "type=keyword", "name", "type=text", "property", "type=nested")
        );
        ensureGreen("idxduplicatehitnames");

        client().prepareIndex("idxduplicatehitnames", "product", "1").setSource(jsonBuilder().startObject()
            .field("name", "product1")
            .array("categories", "1", "2", "3", "4")
            .startArray("property")
            .startObject().field("id", 1).endObject()
            .startObject().field("id", 2).endObject()
            .startObject().field("id", 3).endObject()
            .endArray()
            .endObject()).get();
        client().prepareIndex("idxduplicatehitnames", "product", "2").setSource(jsonBuilder().startObject()
            .field("name", "product2")
            .array("categories", "1", "2")
            .startArray("property")
            .startObject().field("id", 1).endObject()
            .startObject().field("id", 5).endObject()
            .startObject().field("id", 4).endObject()
            .endArray()
            .endObject()).get();
        refresh();

//        SearchResponse response = client().prepareSearch("idxduplicatehitnames").setTypes("product")
//            .addAggregation(terms("category").field("categories").subAggregation(
//                nested("property", "property").subAggregation(
//                    terms("property_id").field("property.id")
//                )
//                .innerHit(new InnerHitBuilder())
//            ))
//            .innerHit(new InnerHitBuilder())
//            .get();

//        SearchResponse response = (SearchResponse)client().prepareSearch(new String[]{"test"})
//            .setQuery(boolQuery()
//                .must(matchQuery("key", "value"))
//                .should(
//                    nestedQuery("nested_type", matchAllQuery(), ScoreMode.None)
//                        .innerHit(new InnerHitBuilder()), ScoreMode.None)
//                .innerHit(new InnerHitBuilder()))
//            .get();

//
//        SearchResponse response = client().prepareSearch("idxduplicatehitnames").addAggregation(nested("to_method", "methods")
//            .subAggregation(filter("num_string_params1",
//                nestedQuery("methods.parameters", termQuery("methods.parameters.type", "String"), ScoreMode.None).innerHit(new InnerHitBuilder("n1"))))
//                .subAggregation(filter("num_string_params2",
//                    nestedQuery("methods.parameters", termQuery("methods.parameters.type", "String"), ScoreMode.None).innerHit(new InnerHitBuilder("n2"))))
//                .subAggregation(filter("num_string_params3",
//                    nestedQuery("methods.parameters", termQuery("methods.parameters.type", "String"), ScoreMode.None).innerHit(new InnerHitBuilder("n1"))))
//        ).get();

//        SearchResponse response = client()
//            .prepareSearch("idxduplicatehitnames")
//            .setTypes("product")
//            .addAggregation(
//                terms("category")
//                    .field("categories")
//                    .subAggregation(filter("n1", nestedQuery("property", termQuery("property.id", 1D), ScoreMode.None).innerHit(new InnerHitBuilder("ih1"))))
//                    .subAggregation(filter("n2", nestedQuery("property", termQuery("property.id", 2D), ScoreMode.None).innerHit(new InnerHitBuilder("ih2"))))
//                    .subAggregation(filter("n3", nestedQuery("property", termQuery("property.id", 5D), ScoreMode.None).innerHit(new InnerHitBuilder("ih1"))))
//            )
//            .get();


        SearchResponse response = client()
            .prepareSearch("idxduplicatehitnames")
            .setTypes("product")
            .addAggregation(
                terms("category")
                    .field("categories")
                    .subAggregation(filter("n1", nestedQuery("property", termQuery("property.id", 1D), ScoreMode.None).innerHit(new InnerHitBuilder("ih1"))))
                    .subAggregation(filter("n2", nestedQuery("property", termQuery("property.id", 2D), ScoreMode.None).innerHit(new InnerHitBuilder("ih2"))))
                    .subAggregation(filter("n3", nestedQuery("property", termQuery("property.id", 5D), ScoreMode.None).innerHit(new InnerHitBuilder("ih1")).innerHit(new InnerHitBuilder("ih1"))))
            )
            .addAggregation(
                terms("category2")
                    .field("categories")
                    .subAggregation(filter("n4", nestedQuery("property", termQuery("property.id", 1D), ScoreMode.None).innerHit(new InnerHitBuilder("ih1"))))
                    .subAggregation(filter("n5", nestedQuery("property", termQuery("property.id", 2D), ScoreMode.None).innerHit(new InnerHitBuilder("ih2"))))
                    .subAggregation(filter("n6", nestedQuery("property", termQuery("property.id", 5D), ScoreMode.None).innerHit(new InnerHitBuilder("ih1")).innerHit(new InnerHitBuilder("ih1"))))
            )
            .get();

        assertNoFailures(response);
//        assertHitCount(response, 2);
    }
```

I've tried a few things but can't seem to get NestedQueryBuilder::extractInnerHitBuilders to be called. I'll keep working on it meanwhile.

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java [hunk 1]
```diff
@@ -460,6 +460,10 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws I
     @Override
     protected void extractInnerHitBuilders(Map<String, InnerHitContextBuilder> innerHits) {
         if (innerHitBuilder != null) {
+            if (innerHits.containsKey(innerHitBuilder.getName())) {
+                throw new IllegalArgumentException("innerHits already contains an entry for key [" + innerHitBuilder.getName() + "]");
```
Comment (jimczi) [inline modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: we should use the same format for the message than the option in the request, can you change `innerHits` to `[inner_hits]`  ?

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasParentQueryBuilder.java [hunk 2]
```diff
@@ -285,6 +285,10 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws I
     @Override
     protected void extractInnerHitBuilders(Map<String, InnerHitContextBuilder> innerHits) {
         if (innerHitBuilder != null) {
+            if (innerHits.containsKey(innerHitBuilder.getName())) {
+                throw new IllegalArgumentException("innerHits already contains an entry for key [" + innerHitBuilder.getName() + "]");
```
Comment (jimczi) [inline modules/parent-join/src/main/java/org/elasticsearch/join/query/HasParentQueryBuilder.java]: Same here, `innerHits`-> `[inner_hits]`

File: server/src/main/java/org/elasticsearch/index/query/NestedQueryBuilder.java [hunk 3]
```diff
@@ -317,10 +318,14 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
     @Override
     public void extractInnerHitBuilders(Map<String, InnerHitContextBuilder> innerHits) {
         if (innerHitBuilder != null) {
+            if (innerHits.containsKey(innerHitBuilder.getName())) {
+                throw new IllegalArgumentException("innerHits already contains an entry for key [" + innerHitBuilder.getName() + "]");
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/NestedQueryBuilder.java]: `innerHits` -> `[inner_hits]`

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java [hunk 4]
```diff
@@ -460,6 +460,10 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws I
     @Override
     protected void extractInnerHitBuilders(Map<String, InnerHitContextBuilder> innerHits) {
         if (innerHitBuilder != null) {
+            if (innerHits.containsKey(innerHitBuilder.getName())) {
```
Comment (jimczi) [inline modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: If `innerHitBuilder.getName()` is null we use the `type` as the name (see below). Can you build the name first and check in the inner hits map with the inferred one ?:
````
String name = innerHitBuilder.getName() != null ? innerHitBuilder.getName() : type;
if (innerHits.containsKey(name)) {
...
````

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasParentQueryBuilder.java [hunk 5]
```diff
@@ -285,6 +285,10 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws I
     @Override
     protected void extractInnerHitBuilders(Map<String, InnerHitContextBuilder> innerHits) {
         if (innerHitBuilder != null) {
+            if (innerHits.containsKey(innerHitBuilder.getName())) {
```
Comment (jimczi) [inline modules/parent-join/src/main/java/org/elasticsearch/join/query/HasParentQueryBuilder.java]: See above, we should check with the inferred name as well if innerHitBuilder.getName is null.

File: server/src/main/java/org/elasticsearch/index/query/NestedQueryBuilder.java [hunk 6]
```diff
@@ -317,10 +318,14 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
     @Override
     public void extractInnerHitBuilders(Map<String, InnerHitContextBuilder> innerHits) {
         if (innerHitBuilder != null) {
+            if (innerHits.containsKey(innerHitBuilder.getName())) {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/index/query/NestedQueryBuilder.java]: here the `path` is used if the name is null, can you infer the name first and then check inside the map ?

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/NestedIT.java [hunk 7]
```diff
@@ -674,4 +679,42 @@ public void testFilterAggInsideNestedAgg() throws Exception {
...
+        );
+        ensureGreen("idxnullhitnames");
+
+        SearchResponse response = client().prepareSearch("idxnullhitnames")
+            .setQuery(boolQuery()
+                .should(nestedQuery("property", termQuery("property.id", 1D), ScoreMode.None).innerHit(new InnerHitBuilder()))
+                .should(nestedQuery("property", termQuery("property.id", 1D), ScoreMode.None).innerHit(new InnerHitBuilder()))
+                .should(nestedQuery("property", termQuery("property.id", 1D), ScoreMode.None).innerHit(new InnerHitBuilder()))
+            ).get();
+        assertNoFailures(response);
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/search/aggregations/bucket/NestedIT.java]: This test should fail, we use the path of the nested field as the name when it is not provided so we should also detect the duplication here.

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks for updating @dvehar . I left more comments.

File: server/src/test/java/org/elasticsearch/search/aggregations/bucket/NestedIT.java (hunk 7 above)
Comment (dvehar) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/NestedIT.java]: Ah ok. When you posted the example [here](https://github.com/elastic/elasticsearch/pull/37645#issuecomment-457146446) I was thinking it should not fail. I'll fix this up.

Comment (jimczi) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/NestedIT.java]: ah ok, sorry it was not clear

Comment (dvehar) [reply server/src/test/java/org/elasticsearch/search/aggregations/bucket/NestedIT.java]: no worries, I think it was just my limited understanding of things.

**Comment by jimczi** [comment]
@elasticmachine test this please

**Comment by jimczi** [comment]
@elasticmachine test this please

**Review by jimczi — COMMENTED** [review]
Thanks @dvehar , the changes look good to me. The tests are failing on unrelated issues so can you merge master in your branch (a simple merge) to make sure that we run the tests with the latest fixes ?

**Comment by dvehar** [comment]
@elasticmachine test this please

**Comment by dvehar** [comment]
@jimczi done

**Comment by jimczi** [comment]
@elasticmachine test this please

**Comment by javanna** [comment]
hi @dvehar that last test that was failing is now addressed. could you merge master in so we can re-run tests and hopefully have a green build? ;)

**Comment by dvehar** [comment]
@javanna done

**Comment by jimczi** [comment]
@elasticmachine test this please

**Comment by javanna** [comment]
retest this please

**Comment by jimczi** [comment]
retest this please

**Comment by jimczi** [comment]
retest this please

**Review by jimczi — APPROVED** [review]
Thanks @dvehar , CI is finally green. Thanks for iterating and sorry for the unrelated build failures ;).

**Comment by dvehar** [comment]
@jimczi Thanks for helping me get it through!

**Pull request merged**